resources/type-system/flow-analysis-v0.md

```diff
@@ -0,0 +1,339 @@
+
+- `false(E)`, where `E` is an expression, represents the *flow model* just after
+  execution of `E`, assuming that `E` completes normally and evaluates to `false`.
+
+- `null(E)`, where `E` is an expression, represents the *flow model* just after
+  execution of `E`, assuming that `E` completes normally and evaluates to `null`.
+
+- `notNull(E)`, where `E` is an expression, represents the *flow model* just
+  after execution of `E`, assuming that `E` completes normally and does not
+  evaluate to `null`.
+
+- `break(S)`, where `S` is a `do`, `for`, `switch`, or `while` statement,
+  represents TODO.
+
+Note that `true`, `false`, `null`, and `notNull` are defined for all expressions
+regardless of their static types.
+
+We also make use of the following auxiliary functions:
+
+- `joinV(VM1, VM2)`, where `VM1` and `VM2` are variable models, represents the
+  union of two variable models, defined as follows:
+  - If `VM1 = VariableModel(d1, p1, s1, a1, u1, c1)` and
+  - If `VM2 = VariableModel(d2, p2, s2, a2, u2, c2)` then
+  - `VM3 = VariableModel(d3, p3, s3, a3, u3, c3)` where
+   - `d3 = d1 = d2`
+     - Note that all models must agree on the declared type of a variable
+   - `p3 = p1 ^ p2`
+     - `p1` and `p2` are totally ordered subsets of a global partial order.
+  Their intersection is a subset of each, and as such is also totally ordered.
+   - `s3 = s1 U s2`
+     - The set of test sites is the union of the test sites on either path
+   - `a3 = a1 && a2`
+     - A variable is definitely assigned in the join of two models iff it is
+       definitely assigned in both.
+   - `u3 = u1 && u2`
+     - A variable is definitely unassigned in the join of two models iff it is
+       definitely unassigned in both.
+   - `c3 = c1 || c2`
+     - A variable is captured in the join of two models iff it is captured in
+       either.
+
+- `split(M)`, where `M = FlowModel(r, VM)` is a flow model which models program
+  nodes inside of a control flow split, and is defined as `FlowModel(r2, VM)`
+  where `r2` is `r` with `true` pushed as the top element of the stack.
+
+- `drop(M)`, where `M = FlowModel(r, VM)` is a flow model which models program
+  nodes after a control flow split which are only reachable by one path through
+  the split, and is defined as `FlowModel(r1, VM)` where `r0 = pop(r)` and `r1 =
+  push(pop(r0), top(r0) && top(r1))`.  Equivalently, `drop(M)` may be thought of
+  as `join(M, M)`.
+
+- `join(M1, M2)`, where `M1` and `M2` are flow models, represents the union of
+  two flow models and is defined as follows:
+
+  We define `join(M1, M2)` where `M1 = FlowModel(r1, VI1)` and `M2 =
+  FlowModel(r2, VI2))` and `pop(r1) = pop(r2) = r0` for some `r0` to be `M3` where:
+    - if `top(r1)` is true and `top(r2)` is false, then `M3` is `FlowModel(r0, VI1)`.
+    - if `r1` is false and `r2` is true, then `M3` is `FlowModel(pop(r2), VI2)`.
+    - otherwise `M3 = FlowModel(r3, VI3)` where:
```

Comment:
I'm not 100% clear whether I'm doing the right thing here.  It's clear what to do if one of the branches is provisionally reachable and the other is not, and it's clear what to do if both are provisionally reachable.  But what about if both are not provisionally reachable?  My guess here is that the right thing to do is to treat them both as provisionally reachable - that is, keep both of their states.  But I'm not 100% sure about that.  

```dart
  foo(Object x, bool b) {
    throw "Unimplemented";
    if (b) {
      if (x is! int) return;
      throw "Unimplemented";
   } else {
     if (x is! int return;
     throw "Unimplemented";
  }
  x.isEven; // should work?
}
```